### PR TITLE
add new primitive atomizeArray

### PR DESCRIPTION
**Pull request by tamrazov**
add new primitive atomizeArray following the example of atomizeSet, there is a question about duplicates in state and method filter, do not merge in until we discuss @artalar.

### CODE REVIEW

**Comment by artalar** [comment]
Please, implement this type

```ts
export type ArrayAtom<T> = WithReducers<
  AtomMut<Array<T>>,
  {
    add(state: Array<T>, value: T): Array<T>
    remove(
      state: Array<T>,
      from: number,
      to?: number /* = from + 1 */,
    ): Array<T>
    replace(
      state: Array<T>,
      newStatePart: Array<T>,
      from?: number /* = 0 */,
      to?: number /* = this.length - 1 */,
    ): Array<T>
    clear(): Array<T>
    reset(): Array<T>
  }
>
```

**Comment by artalar** [comment]
btw, all kind of transformations could be described by `flatMap` (and `[...state, added]`):

```ts
// remove from a middle
listAtom(ctx, state => state.flatMap((v) => (v === 2 ? [] : v)))
// [1, 3]

// add to a middle
listAtom(ctx, state => state.flatMap((v) => (v === 2 ? [2, 2.5] : v)))
// [1, 2, 2.5, 3]
```

do we really need extra methods? 🤔

**Comment by tamrazov** [comment]
@artalar, The problem is that I don't get how write tests for atomizeArray. Please, describe one case for example method `add` for me in this conversation. The following tests I will describe myself.

**Comment by artalar** [comment]
Hi! I was thinking about all of it and propose a small changes:

```ts
export type ArrayAtom<T> = WithReducers<
  AtomMut<Array<T>>,
  {
    put(
      state: Array<T>,
      value: T,
      position?:
        | number
        | ((el: T, index: number) => boolean) /* = this.length */
    ): Array<T>;
    remove(
      state: Array<T>,
      from: number,
      to?: number /* = from */
    ): Array<T>;
    replace(
      state: Array<T>,
      newStatePart: Array<T>,
      from?: number /* = 0 */,
      to?: number /* = this.length - 1 */
    ): Array<T>;
    clear(): Array<T>;
    reset(): Array<T>;
  }
>;

// EXAMPLES

atomizeArray([1, 2, 3]).put(ctx, 4)
// [1, 2, 3, 4]

atomizeArray([1, 2, 3]).put(ctx, 4, 0)
// [4, 2, 3]

atomizeArray([1, 2, 3]).put(ctx, 4, (el) => el === 2)
// [1, 4, 3]
```

**Comment by artalar** [comment]
@altamraz please, change `to` to `count` and allow callback for `from`:

> In this case `remove` will super simple: `(state, from, count = 1) => replace(state, [], from, count)`

```ts
export type ArrayAtom<T> = WithReducers<
  AtomMut<Array<T>>,
  {
    put(
      state: Array<T>,
      value: T,
      position?:
        | number
        | ((el: T, index: number) => boolean) /* = this.length */
    ): Array<T>;
    remove(
      state: Array<T>,
      from: number | ((el: T, index: number) => boolean),
      count?: number /* = 1 */
    ): Array<T>;
    replace(
      state: Array<T>,
      newStatePart: Array<T>,
      from?: number | ((el: T, index: number) => boolean) /* = 0 */,
      count?: number /* = this.length - from */
    ): Array<T>;
    clear(): Array<T>;
    reset(): Array<T>;
  }
>;
```

**Comment by tamrazov** [comment]
Add new Type ArrayAtom
```
export type ArrayAtom<T> = WithReducers<
  AtomMut<Array<T>>,
  {
    toReversed(): Array<T>
    toSorted(compareFn?: (a: T, b: T) => number): Array<T>
    toSpliced(state: Array<T>, start: number, deleteCount: number, ...items: T[]): Array<T>
    with(state: Array<T>, index: number, value: T): Array<T>
  }
>
```

File: packages/primitives/src/atomizeArray.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      toReversed: (state) => {
+        const newState = state.slice(0)
+ 
+        return newState.reverse()
+      },
+      toSorted: (state, compareFn) => {
+        const newState = state.slice(0)
+ 
+        return newState.sort(compareFn)
+      },
```
Comment (artalar) [inline packages/primitives/src/atomizeArray.ts]:
```suggestion
      toReversed: (state) => state.slice(0).reverse(),
      toSorted: (state, compareFn) => state.slice(0).sort(compareFn),
```

File: packages/primitives/src/atomizeArray.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      toSpliced: (state, start, deleteCount, ...items) => {
+        const newState = state.slice(0)
```
Comment (artalar) [inline packages/primitives/src/atomizeArray.ts]: reuse `state` variable

File: packages/primitives/src/atomizeArray.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        newState.splice(start, deleteCount, ...items)
+ 
+        return newState
+      },
+      with: (state, index, value) => {
+        const newState = state.slice(0)
```
Comment (artalar) [inline packages/primitives/src/atomizeArray.ts]: reuse `state` variable

**Review by artalar — CHANGES_REQUESTED** [review]

**Review by artalar — APPROVED** [review]

**Pull request merged**